src/block-data-providers/esplora.provider.ts

```diff
@@ -162,33 +162,59 @@ export class EsploraProvider
     }
 
     private async processBlock(height: number, hash: string) {
+        const state = await this.getState();
+        const lastProcessedTxIndex = state?.lastProcessedTxIndex ?? 0;
+
         const txids = await this.getTxidsForBlock(hash);
 
-        for (let i = 1; i < txids.length; i += this.batchSize) {
+        for (
+            let i = lastProcessedTxIndex;
+            i < txids.length;
+            i += this.batchSize
+        ) {
             const batch = txids.slice(
                 i,
                 Math.min(i + this.batchSize, txids.length),
             );
 
-            await Promise.all(
-                batch.map(async (txid) => {
-                    const tx = await this.getTx(txid);
-                    const vin: TransactionInput[] = tx.vin.map((input) => ({
-                        txid: input.txid,
-                        vout: input.vout,
-                        scriptSig: input.scriptsig,
-                        prevOutScript: input.prevout.scriptpubkey,
-                        witness: input.witness,
-                    }));
-                    const vout = tx.vout.map((output) => ({
-                        scriptPubKey: output.scriptpubkey,
-                        value: output.value,
-                    }));
+            try {
+                await Promise.all(
+                    batch.map(async (txid) => {
+                        const tx = await this.getTx(txid);
+                        const vin: TransactionInput[] = tx.vin.map((input) => ({
+                            txid: input.txid,
+                            vout: input.vout,
+                            scriptSig: input.scriptsig,
+                            prevOutScript: input.prevout.scriptpubkey,
+                            witness: input.witness,
+                        }));
+                        const vout = tx.vout.map((output) => ({
+                            scriptPubKey: output.scriptpubkey,
+                            value: output.value,
+                        }));
 
-                    await this.indexTransaction(txid, vin, vout, height, hash);
-                }, this),
-            );
+                        await this.indexTransaction(
+                            txid,
+                            vin,
+                            vout,
+                            height,
+                            hash,
+                        );
+                    }, this),
+                );
+
+                state.lastProcessedTxIndex = i + this.batchSize - 1;
+                await this.setState(state);
+            } catch (error) {
+                this.logger.error(
+                    `Error processing transactions in block at height ${height}, hash ${hash}: ${error.message}`,
+                );
+                throw error;
+            }
         }
+
+        state.lastProcessedTxIndex = 0;
+        await this.setState(state);
```

Comment:
I don't get why we are setting it to 0 here. We should do anything here.

Reply:
once we are done processing the entire block. Reset the last processed transaction index .

Reply:
why do we need to reset after finishing the processing?